### PR TITLE
Save home_dashboard preference in preferences table for a user

### PR DESCRIPTION
**Pull request by utkarshcmu**
Some more work done for preferences feature. Now a user can save home_dashboard preference for each org.

PS: This PR is for preferences branch.

### CODE REVIEW

File: public/app/features/dashboard/dashnav/dashnav.ts [hunk 1]
```diff
@@ -168,6 +168,27 @@ export class DashNavCtrl {
       });
     };
 
+    $scope.saveDashboardAsHome = function() {
+      backendSrv.get('/api/org').then(function(org) {
```
Comment (torkelo) [inline public/app/features/dashboard/dashnav/dashnav.ts]: you have current org on contextSrv

**Comment by utkarshcmu** [comment]
Done, Thanks !!!

**Comment by torkelo** [comment]
one thing that I am not sure about is the structure that we have now is enough

```
id	pref_id	pref_type	preferences
1	4	user	{"timezone":"utc", "home_dashboard":"cpu-summary"}
2	3	user	{"timezone":"utc", "timepicker":{"from":"now-3h","to":"now"}}
3	1	org	{"home_dashboard":"app-summary","refresh_every":"never"}
4	0	global	{"home_dashboard":"company-summary","timepicker":{"from":"now-1h","to":"now"}}
```

pref_type should maybe be level? The above structure does not support user options per org, like home dashboard. not sure I like have the orgId in the user preferences like i this PR. 

Let me think for a bit, will get back

**Comment by utkarshcmu** [comment]
Agreed, even I also thought a lot about this structure. It is a bit messy, Yes.

I will also think over for a better relational structure and get back to u.

**Comment by torkelo** [comment]
Maybe something more straight forward like this?

| id  | org_id      | user_id        | home_dashboard_id | timezone | theme |
|-----|-------------|----------------|---------------------------|---------------|------------|-----------------|
| 1    | 4            | 6          | 22 | 'utc' | 'dark'

If org_id is null settings are for user (and users all orgs), if org_id and user_id the settings are active 
for that org and user, if user_id is null and org_id has value it's defaults for that org. 

The only thing that is going to be tricky is the api and internal commands, some settings (like theme) should not be stored for specific org (share setting for a user's all orgs, but should be able to set default on org level).

**Comment by utkarshcmu** [comment]
I would like to suggest a slight change in your table because it will be hard to scale for each kind of preference. Today we may be able to think of may be 3-4 preferences, but users will come up with way more than these and then we have to add one column per preference, which i do not think is a generic solution.

How about this small change in your table?

| id | org_id | user_id | preference |
|----|----------|------------|---------------|
|1|1|1|{"home_dashboard_id":3}|
|2|0|1|{"home_dashboard_id":5,"theme":"dark"}|
|3|1|0|{"timezone":"utc","theme":"dark"}|
|4|0|0|{"home_dashboard_id":7}|

org_id > 0 , user_id > 0 => user preference for that org
org_id = 0 , user_id > 0 => user preference for all orgs
org_id > 0 , user_id = 0 => org preference for all
org_id = 0 , user_id = 0 => global grafana preference

What do you think about this?

**Comment by torkelo** [comment]
that can work, need a version column (to handle optimistic concurrency)  and a created/updated column

**Comment by utkarshcmu** [comment]
Will update the schema of preferences table asap. Thanks for the feedback!!!

Closing this PR as this is not valid anymore according to the new schema.

**Pull request closed without merging**